File: .skills/openclaw-skills/skills/0xnyk/xint/lib/costs.test.ts
/**
 * Unit tests for costs.ts
 * Run with: bun test lib/costs.test.ts
 */

import { describe, test, expect, beforeEach } from "bun:test";
import { trackCost, trackCostDirect, checkBudget, getTodayCosts, setBudget, getCostSummary } from "./costs";

// Test file path - use temp for tests
const TEST_DATA_FILE = "/tmp/xint-test-costs.json";

describe("Cost Tracking", () => {
  beforeEach(() => {
    // Reset for each test
  });

  test("trackCost calculates correct cost for search", () => {
    const entry = trackCost("search", "/2/tweets/search/recent", 10);
    
    expect(entry.operation).toBe("search");
    expect(entry.endpoint).toBe("/2/tweets/search/recent");
    expect(entry.tweets_read).toBe(10);
    expect(entry.cost_usd).toBe(0.05); // 0.005 * 10
  });

  test("trackCost calculates correct cost for profile", () => {
    const entry = trackCost("profile", "/2/users/by/username/test", 5);
    
    expect(entry.cost_usd).toBe(0.025); // 0.005 * 5
  });

  test("trackCost calculates correct cost for like action", () => {
    const entry = trackCost("like", "/2/users/me/likes", 0);
    
    expect(entry.cost_usd).toBe(0.01); // per_call rate
  });

  test("trackCost calculates correct cost for trends", () => {
    const entry = trackCost("trends", "/2/trends/by/woeid/1", 0);
    
    expect(entry.cost_usd).toBe(0.10); // per_call rate
  });

  test("trackCost uses fallback rate for unknown operation", () => {
    const entry = trackCost("unknown_op", "/unknown", 20);
    
    // Should use default 0.005 per tweet
    expect(entry.cost_usd).toBe(0.1); // 0.005 * 20
  });

  test("checkBudget returns correct status", () => {
    const status = checkBudget();
    
    expect(status).toHaveProperty("allowed");
    expect(status).toHaveProperty("spent");
    expect(status).toHaveProperty("limit");
    expect(status).toHaveProperty("remaining");
    expect(status).toHaveProperty("warning");
  });

  test("getTodayCosts returns today aggregate", () => {
    const today = getTodayCosts();
    
    expect(today).toHaveProperty("date");
    expect(today).toHaveProperty("total_cost");
    expect(today).toHaveProperty("calls");
    expect(today).toHaveProperty("tweets_read");
    expect(today).toHaveProperty("by_operation");
  });

  test("setBudget updates daily limit", () => {
    const originalLimit = checkBudget().limit;
    
    setBudget(5.0);
    const status = checkBudget();
    
    expect(status.limit).toBe(5.0);
    
    // Restore original
    setBudget(originalLimit);
  });

  test("getCostSummary returns string for today", () => {
    const summary = getCostSummary("today");
    
    expect(typeof summary).toBe("string");
    expect(summary.length).toBeGreaterThan(0);
  });

  test("getCostSummary returns string for week", () => {
    const summary = getCostSummary("week");
    
    expect(typeof summary).toBe("string");
  });

  test("getCostSummary returns string for month", () => {
    const summary = getCostSummary("month");
    
    expect(typeof summary).toBe("string");
  });

  test("getCostSummary returns string for all", () => {
    const summary = getCostSummary("all");
    
    expect(typeof summary).toBe("string");
  });
});

describe("Direct Cost Tracking (xAI/Grok)", () => {
  test("trackCostDirect records exact USD amount", () => {
    const entry = trackCostDirect("grok_chat", "https://api.x.ai/v1/chat/completions", 0.0042);

    expect(entry.operation).toBe("grok_chat");
    expect(entry.cost_usd).toBe(0.0042);
    expect(entry.tweets_read).toBe(0);
  });

  test("trackCostDirect rounds to avoid floating-point noise", () => {
    const entry = trackCostDirect("xai_article", "https://api.x.ai/v1/responses", 0.00000123);

    expect(entry.cost_usd).toBe(0.000001);
  });
});

describe("Cost Rates", () => {
  test("COST_RATES has search rate", () => {
    const { COST_RATES } = require("./costs");
    
    expect(COST_RATES.search.per_tweet).toBe(0.005);
    expect(COST_RATES.search.per_call).toBe(0);
  });

  test("COST_RATES has like rate", () => {
    const { COST_RATES } = require("./costs");
    
    expect(COST_RATES.like.per_call).toBe(0.01);
  });

  test("COST_RATES has follow/unfollow rates", () => {
    const { COST_RATES } = require("./costs");

    expect(COST_RATES.follow.per_call).toBe(0.01);
    expect(COST_RATES.unfollow.per_call).toBe(0.01);
  });

  test("COST_RATES has stream rates", () => {
    const { COST_RATES } = require("./costs");

    expect(COST_RATES.media_metadata.per_tweet).toBe(0.005);
    expect(COST_RATES.stream_connect.per_tweet).toBe(0.005);
    expect(COST_RATES.stream_rules_list.per_call).toBe(0.01);
    expect(COST_RATES.stream_rules_add.per_call).toBe(0.01);
    expect(COST_RATES.stream_rules_delete.per_call).toBe(0.01);
  });

  test("COST_RATES has trends rate", () => {
    const { COST_RATES } = require("./costs");
    
    expect(COST_RATES.trends.per_call).toBe(0.10);
  });

  test("COST_RATES has bookmarks rate", () => {
    const { COST_RATES } = require("./costs");

    expect(COST_RATES.bookmarks.per_tweet).toBe(0.005);
  });

  test("COST_RATES has xAI/Grok rates", () => {
    const { COST_RATES } = require("./costs");

    expect(COST_RATES.grok_chat.per_call).toBe(0.0005);
    expect(COST_RATES.grok_vision.per_call).toBe(0.005);
    expect(COST_RATES.xai_article.per_call).toBe(0.0015);
    expect(COST_RATES.xai_x_search.per_call).toBe(0.001);
  });
});

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/costs_test.rs
#[cfg(test)]
mod tests {
    use crate::costs::{cost_rate, track_cost, track_cost_direct, check_budget, set_budget, get_cost_summary};
    use std::path::PathBuf;
    use std::fs;

    fn test_data_path() -> PathBuf {
        let path = PathBuf::from("/tmp/xint-rs-test-costs.json");
        let _ = fs::remove_file(&path);
        path
    }

    #[test]
    fn test_cost_rate_search() {
        let (per_tweet, per_call) = cost_rate("search");
        assert_eq!(per_tweet, 0.005);
        assert_eq!(per_call, 0.0);
    }

    #[test]
    fn test_cost_rate_like() {
        let (per_tweet, per_call) = cost_rate("like");
        assert_eq!(per_tweet, 0.0);
        assert_eq!(per_call, 0.01);
    }

    #[test]
    fn test_cost_rate_trends() {
        let (per_tweet, per_call) = cost_rate("trends");
        assert_eq!(per_tweet, 0.0);
        assert_eq!(per_call, 0.10);
    }

    #[test]
    fn test_cost_rate_unknown() {
        let (per_tweet, per_call) = cost_rate("unknown_operation");
        assert_eq!(per_tweet, 0.005); // fallback
        assert_eq!(per_call, 0.0);
    }

    #[test]
    fn test_cost_rate_grok() {
        let (per_tweet, per_call) = cost_rate("grok_chat");
        assert_eq!(per_tweet, 0.0);
        assert_eq!(per_call, 0.001);

        let (per_tweet, per_call) = cost_rate("grok_vision");
        assert_eq!(per_tweet, 0.0);
        assert_eq!(per_call, 0.005);

        let (per_tweet, per_call) = cost_rate("xai_article");
        assert_eq!(per_tweet, 0.0);
        assert_eq!(per_call, 0.003);
    }

    #[test]
    fn test_track_cost_direct() {
        let path = test_data_path();

        let entry = track_cost_direct(&path, "grok_chat", "https://api.x.ai/v1/chat/completions", 0.0042);

        assert_eq!(entry.operation, "grok_chat");
        assert_eq!(entry.tweets_read, 0);
        assert!((entry.cost_usd - 0.0042).abs() < 0.000001);

        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_track_cost() {
        let path = test_data_path();
        
        let entry = track_cost(&path, "search", "/2/tweets/search/recent", 10);
        
        assert_eq!(entry.operation, "search");
        assert_eq!(entry.tweets_read, 10);
        assert!((entry.cost_usd - 0.05).abs() < 0.001);
        
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_check_budget() {
        let path = test_data_path();
        
        let status = check_budget(&path);
        
        assert!(status.limit > 0.0);
        assert!(status.spent >= 0.0);
        
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_set_budget() {
        let path = test_data_path();
        
        set_budget(&path, 5.0);
        let status = check_budget(&path);
        
        assert_eq!(status.limit, 5.0);
        
        // Restore default
        set_budget(&path, 1.0);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_get_cost_summary_today() {
        let path = test_data_path();
        
        // Track some costs first
        let _ = track_cost(&path, "search", "/test", 10);
        
        let summary = get_cost_summary(&path, "today");
        
        assert!(summary.contains("API Costs"));
        assert!(summary.contains("Today"));
        
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_get_cost_summary_week() {
        let path = test_data_path();
        
        let summary = get_cost_summary(&path, "week");
        
        assert!(summary.contains("API Costs"));
        
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_get_cost_summary_month() {
        let path = test_data_path();
        
        let summary = get_cost_summary(&path, "month");
        
        assert!(summary.contains("API Costs"));
        
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_get_cost_summary_all() {
        let path = test_data_path();
        
        let summary = get_cost_summary(&path, "all");
        
        assert!(summary.contains("API Costs"));
        assert!(summary.contains("All Time"));
        
        let _ = fs::remove_file(&path);
    }
}
